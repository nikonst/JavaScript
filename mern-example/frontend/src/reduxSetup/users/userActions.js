import { ADD_USER, DELETE_USER } from './userTypes'

export const addUser = () => {
    return {
        type: ADD_USER
    }
}

export const  deleteUser= () => {
    return {
        type: DELETE_USER
    }
}