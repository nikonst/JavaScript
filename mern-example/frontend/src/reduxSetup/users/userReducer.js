import { v4 as uuidv4 } from 'uuid'
import { GET_USERS, ADD_USER, DELETE_USER, LOADING_USERS } from './userTypes'

const initialState = {
    users: [
        /* { id: uuidv4(), name: "Ian" },
        { id: uuidv4(), name: "Tom" },
        { id: uuidv4(), name: "Ann" }, */
    ],
    loading: false
}

const userReducer = (state = initialState, action) => {
    switch (action.type) {
        case GET_USERS: {
            //console.log("action.payload", action.payload)
            return {
                ...state,
                users: action.payload,
                loading: false
            }
        }
        case ADD_USER: {
            return {
                ...state,
                users: [...state.users, { id: uuidv4(), name: action.payload }]
            }
        }
        case DELETE_USER: {
            console.log("STATE ", state, "ACTION ", action)
            return {
                ...state,
                users: state.users.filter(user => user.id !== action.payload)
            }
        }
        case LOADING_USERS: {
            return {
                ...state,
                loading: true
            }
        }
        default: return state
    }
}

export default userReducer