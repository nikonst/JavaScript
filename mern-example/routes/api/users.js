const express = require('express')
const bcrypt = require('bcryptjs')
const router = express.Router()
const configs = require('./../../config.js')
const jwt = require('jsonwebtoken')

//User model
const User = require('../../models/User')

//POST users/add
//@Access Public
router.post('/add', (req, res) => {
    //res.send("Register")
    const { name, email, password } = req.body
    if( !email || !name || !password) {
        return res.status(400).json({msg: "Enter all fields..."})
    }
    //Check for existing user
    User.findOne({email: email})
    .then(user => {
        if(user) return res.status(400).json({msg: "User allready exists!"})
        const newUser = new User({
            name, email, password
        })
        //Create hash
        bcrypt.genSalt(10, (err, salt) => {
            bcrypt.hash(newUser.password, salt, (err, hash) => {
                if(err) throw err
                newUser.password = hash
                newUser.save()
                .then(user => {
                    jwt.sign(
                        {id: user.id},
                        configs.JWTSECRET,
                        {expiresIn: 3600},
                        (err, token) => {
                            if(err) throw err
                            res.json({
                                token: token,
                                user: {
                                    id: user.id,
                                    name: user.name,
                                    email: user.email
                                }
                            })
                        }
                    )
                    /* res.json({
                        id: user.id,
                        name: user.name,
                        email: user.email
                    }) */
                //.catch(err => console.log(err))
                })
            })
        })
    })
})

module.exports = router