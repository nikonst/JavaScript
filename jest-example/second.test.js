import users from './data/data.json'

test('the data are coreect', () => {
    expect(users).toMatchSnapshot()
    expect(users).toHaveLength(28)
})

users.forEach(element => {
    test('element should have properties id, name, surname, email, country', () => {
        expect(element).toHaveProperty('id')
        expect(element).toHaveProperty('name')
        expect(element).toHaveProperty('surname')
        expect(element).toHaveProperty('email')
        expect(element).toHaveProperty('country')
    })
    
});