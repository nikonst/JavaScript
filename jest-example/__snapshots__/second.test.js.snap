// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the data are coreect 1`] = `
Array [
  Object {
    "country": "Brazil",
    "email": "ac.turpis.egestas@loremloremluctus.edu",
    "id": "16070105-1997",
    "name": "Allegra",
    "surname": "Rowland",
  },
  Object {
    "country": "Israel",
    "email": "adipiscing@consequatdolor.com",
    "id": "16180818-3360",
    "name": "Rae",
    "surname": "Galloway",
  },
  Object {
    "country": "Hungary",
    "email": "ac.ipsum.Phasellus@nostraper.org",
    "id": "16721228-0692",
    "name": "Hu",
    "surname": "Bryan",
  },
  Object {
    "country": "Guinea-Bissau",
    "email": "consectetuer@incursus.edu",
    "id": "16300714-1769",
    "name": "Zenia",
    "surname": "Sampson",
  },
  Object {
    "country": "Jersey",
    "email": "Nullam.nisl.Maecenas@gravidamaurisut.ca",
    "id": "16581024-6727",
    "name": "Caldwell",
    "surname": "Davis",
  },
  Object {
    "country": "Kazakhstan",
    "email": "ultricies.adipiscing.enim@Fusce.org",
    "id": "16600504-3036",
    "name": "Abel",
    "surname": "Osborne",
  },
  Object {
    "country": "Burkina Faso",
    "email": "non@Donec.com",
    "id": "16690928-4918",
    "name": "Mannix",
    "surname": "Franklin",
  },
  Object {
    "country": "Mauritania",
    "email": "justo.eu.arcu@velit.edu",
    "id": "16270519-5770",
    "name": "Shana",
    "surname": "Wilkinson",
  },
  Object {
    "country": "Czech Republic",
    "email": "velit@faucibusut.net",
    "id": "16000406-4291",
    "name": "Owen",
    "surname": "Vincent",
  },
  Object {
    "country": "Uzbekistan",
    "email": "elementum@Aliquamauctor.org",
    "id": "16160625-2334",
    "name": "Whoopi",
    "surname": "Floyd",
  },
  Object {
    "country": "Virgin Islands, United States",
    "email": "gravida.mauris@Inmipede.edu",
    "id": "16750910-2583",
    "name": "Ori",
    "surname": "White",
  },
  Object {
    "country": "Albania",
    "email": "diam.vel.arcu@non.net",
    "id": "16050620-4197",
    "name": "Harrison",
    "surname": "Avila",
  },
  Object {
    "country": "Uganda",
    "email": "et.eros.Proin@idrisus.org",
    "id": "16101218-4766",
    "name": "Curran",
    "surname": "Sweet",
  },
  Object {
    "country": "Mongolia",
    "email": "non.vestibulum.nec@vitae.co.uk",
    "id": "16340112-4627",
    "name": "Bert",
    "surname": "House",
  },
  Object {
    "country": "Wallis and Futuna",
    "email": "vel.sapien.imperdiet@quisdiam.com",
    "id": "16980424-5943",
    "name": "Dean",
    "surname": "Cantu",
  },
  Object {
    "country": "Algeria",
    "email": "Nunc.ullamcorper.velit@sedturpisnec.org",
    "id": "16370621-4594",
    "name": "Alexander",
    "surname": "Crawford",
  },
  Object {
    "country": "Laos",
    "email": "magna@odiovelest.org",
    "id": "16930727-4531",
    "name": "Malachi",
    "surname": "Knapp",
  },
  Object {
    "country": "Ecuador",
    "email": "Pellentesque.habitant.morbi@euismodest.ca",
    "id": "16480427-8705",
    "name": "Hedda",
    "surname": "Haley",
  },
  Object {
    "country": "Angola",
    "email": "egestas.a@atvelitPellentesque.org",
    "id": "16010413-1388",
    "name": "Ashely",
    "surname": "Farley",
  },
  Object {
    "country": "Botswana",
    "email": "vehicula.aliquet@lacusvariuset.com",
    "id": "16160216-2008",
    "name": "Gannon",
    "surname": "Gillespie",
  },
  Object {
    "country": "Central African Republic",
    "email": "pede.ac.urna@auctor.co.uk",
    "id": "16040113-3046",
    "name": "Halla",
    "surname": "Kelley",
  },
  Object {
    "country": "Trinidad and Tobago",
    "email": "Duis@aliquet.co.uk",
    "id": "16150205-2150",
    "name": "Brynne",
    "surname": "Castillo",
  },
  Object {
    "country": "Nauru",
    "email": "blandit.congue.In@Integereu.co.uk",
    "id": "16650123-9674",
    "name": "Kerry",
    "surname": "Delacruz",
  },
  Object {
    "country": "Israel",
    "email": "mollis.nec@pedeSuspendissedui.org",
    "id": "16831020-8825",
    "name": "Aiko",
    "surname": "Jacobson",
  },
  Object {
    "country": "Cambodia",
    "email": "consequat.enim@mus.net",
    "id": "16020202-8866",
    "name": "Carissa",
    "surname": "Jimenez",
  },
  Object {
    "country": "Seychelles",
    "email": "auctor.odio@tinciduntduiaugue.co.uk",
    "id": "16441116-4272",
    "name": "Bradley",
    "surname": "Crawford",
  },
  Object {
    "country": "New Caledonia",
    "email": "aliquam.eros.turpis@enim.org",
    "id": "16020111-4683",
    "name": "Murphy",
    "surname": "Delgado",
  },
  Object {
    "country": "Barbados",
    "email": "neque@Maurisblandit.co.uk",
    "id": "16811227-1468",
    "name": "Madaline",
    "surname": "Levine",
  },
]
`;
